source/types.ts

```diff
@@ -1,274 +1,121 @@
 }
 
-export type RetryFunction = (retryObject: RetryObject) => number;
-
+export type GotReturn = Request | CancelableRequest;
 export type HandlerFunction = <T extends GotReturn>(options: NormalizedOptions, next: (options: NormalizedOptions) => T) => T | Promise<T>;
 
-export interface DefaultRetryOptions {
-	limit: number;
-	methods: Method[];
-	statusCodes: number[];
-	errorCodes: string[];
-	calculateDelay: RetryFunction;
-	maxRetryAfter?: number;
-}
-
-export interface RetryOptions extends Partial<DefaultRetryOptions> {
-	retries?: number;
-}
-
-export type RequestFunction = typeof http.request;
-
-export interface AgentByProtocol {
-	http?: http.Agent;
-	https?: https.Agent;
-}
-
-export interface Delays {
-	lookup?: number;
-	connect?: number;
-	secureConnect?: number;
-	socket?: number;
-	response?: number;
-	send?: number;
-	request?: number;
-}
-
-export type Headers = Record<string, string | string[] | undefined>;
-
-interface ToughCookieJar {
-	getCookieString(currentUrl: string, options: {[key: string]: unknown}, cb: (err: Error | null, cookies: string) => void): void;
-	getCookieString(url: string, callback: (error: Error | null, cookieHeader: string) => void): void;
-	setCookie(cookieOrString: unknown, currentUrl: string, options: {[key: string]: unknown}, cb: (err: Error | null, cookie: unknown) => void): void;
-	setCookie(rawCookie: string, url: string, callback: (error: Error | null, result: unknown) => void): void;
-}
-
-interface PromiseCookieJar {
-	getCookieString(url: string): Promise<string>;
-	setCookie(rawCookie: string, url: string): Promise<unknown>;
+export interface ExtendOptions extends Options {
+	handlers?: HandlerFunction[];
+	mutableDefaults?: boolean;
 }
 
-export const requestSymbol = Symbol('request');
+type Except<ObjectType, KeysType extends keyof ObjectType> = Pick<ObjectType, Exclude<keyof ObjectType, KeysType>>;
 
-/* eslint-disable @typescript-eslint/indent */
-export type DefaultOptions = Merge<
-	Required<
-		Except<
-			GotOptions,
-			// Override
-			'hooks' |
-			'retry' |
-			'timeout' |
-			'context' |
-			'_pagination' |
+export type OptionsOfTextResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType?: 'text'};
+export type OptionsOfJSONResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'json'};
+export type OptionsOfBufferResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'buffer'};
+export type StrictOptions = Except<Options, 'isStream' | 'responseType' | 'resolveBodyOnly'>;
+type ResponseBodyOnly = {resolveBodyOnly: true};
 
-			// Should not be present
-			'agent' |
-			'body' |
-			'cookieJar' |
-			'encoding' |
-			'form' |
-			'json' |
-			'lookup' |
-			'request' |
-			'url' |
-			typeof requestSymbol
-		>
-	>,
-	{
-		hooks: Required<Hooks>;
-		retry: DefaultRetryOptions;
-		timeout: Delays;
-		context: {[key: string]: any};
-		_pagination?: PaginationOptions<unknown>['_pagination'];
```

Comment:
We should make it public in Got 12 if there are no people giving feedback rn.

Reply:
Yup. Feel free to make it public now.